internal/signer/darwin/keychain/keychain.go

```diff
@@ -405,3 +439,91 @@ func certIn(xc *x509.Certificate, xcs []*x509.Certificate) bool {
 	}
 	return false
 }
+
+func (k *Key) EncryptRSA(hashInput hash.Hash, random io.Reader, msg []byte) ([]byte, error) {
+	pub := k.Public()
+	rsaPubKey := pub.(*rsa.PublicKey)
+	return rsa.EncryptOAEP(hashInput, random, rsaPubKey, msg, nil)
+}
+
+func (k *Key) getRSAEncryptAlgor() (C.SecKeyAlgorithm, error) {
+	var algorithms map[crypto.Hash]C.CFStringRef
+	switch pub := k.Public().(type) {
+	case *rsa.PublicKey:
+		if C.SecKeyIsAlgorithmSupported(k.publicKeyRef, C.kSecKeyOperationTypeEncrypt, C.kSecKeyAlgorithmRSASignatureDigestPSSSHA256) == 1 {
+			algorithms = rsaPSSAlgorithms
+		} else if C.SecKeyIsAlgorithmSupported(k.publicKeyRef, C.kSecKeyOperationTypeEncrypt, C.kSecKeyAlgorithmRSAEncryptionOAEPSHA256) == 1 {
+			algorithms = rsaOAEPAlgorithms
+		} else if C.SecKeyIsAlgorithmSupported(k.publicKeyRef, C.kSecKeyOperationTypeEncrypt, C.kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256) == 1 {
+			algorithms = rsaPKCS1v15Algorithms
+		}
+	default:
+		return 0, fmt.Errorf("unsupported algorithm %T", pub)
+	}
+	return algorithms[k.hashFunc], nil
+}
+
+func (k *Key) getEncryptAlgorithm() (C.SecKeyAlgorithm, error) {
+	if k.hashFunc == 0 {
+		k.hashFunc = crypto.SHA256
+	}
+	return k.getRSAEncryptAlgor()
+}
+
+func (k *Key) getRSADecryptAlgor() (C.SecKeyAlgorithm, error) {
+	var algorithms map[crypto.Hash]C.CFStringRef
+	switch pub := k.Public().(type) {
+	case *rsa.PublicKey:
+		if C.SecKeyIsAlgorithmSupported(k.publicKeyRef, C.kSecKeyOperationTypeDecrypt, C.kSecKeyAlgorithmRSASignatureDigestPSSSHA256) == 1 {
+			algorithms = rsaPSSAlgorithms
+		} else if C.SecKeyIsAlgorithmSupported(k.publicKeyRef, C.kSecKeyOperationTypeDecrypt, C.kSecKeyAlgorithmRSAEncryptionOAEPSHA256) == 1 {
+			algorithms = rsaOAEPAlgorithms
+		} else if C.SecKeyIsAlgorithmSupported(k.publicKeyRef, C.kSecKeyOperationTypeDecrypt, C.kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256) == 1 {
+			algorithms = rsaPKCS1v15Algorithms
+		}
+	default:
+		return 0, fmt.Errorf("unsupported algorithm %T", pub)
```

Comment:
Let's create a constant named UNKNOWN_SECKEY_ALGORITHM, initialize it to 0, and return it if we can't determine the algorithm.

Reply:
Similar comment for the error message

Reply:
Keychain.go's author wrote the "unsupported algorithm %T" error message in the Sign() which I then reused. Should I still change this error message or leave it the same for code consistency?

Reply:
Let's update your code. We can always contribute detailed logs later!